packages/workspace/src/browser/workspace-action-provider.ts

```diff
@@ -0,0 +1,56 @@
+/********************************************************************************
+ * Copyright (C) 2019 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import { QuickOpenBaseAction, QuickOpenItem, QuickOpenActionProvider, QuickOpenAction } from '@theia/core/lib/browser';
+import { WorkspaceServer } from '../common';
+
+@injectable()
+export class RemoveWorkspaceAction extends QuickOpenBaseAction {
+
+    @inject(WorkspaceServer)
+    protected readonly workspaceServer: WorkspaceServer;
+
+    constructor() {
+        super({ id: 'remove:workspace:action' });
+        this.class = 'fa fa-times';
+    }
+
+    async run(item?: QuickOpenItem): Promise<void> {
+        if (item && item.getUri()) {
+            return (
+                this.workspaceServer.removeRecentWorkspace &&
```

Comment:
Should not be action unavailable at all if `removeRecentWorkspace` is not implemented, i.e. filtering should be done in `WorkspaceActionProvider`.

Reply:
I think perhaps I should keep the original check, but also perform the check like you mentioned for the `WorkspaceActionProvider`. Having the logic only in the `WorkspaceActionProvider` does not guarantee that another action-provider will use the `RemoveWorkspaceAction` correctly. 

What do you think, should we be a little defensive and check anyways that the method is in fact implemented?